src/Symfony/Component/JsonEncoder/Mapping/Decode/AttributePropertyMetadataLoader.php

```diff
@@ -0,0 +1,107 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\JsonEncoder\Mapping\Decode;
+
+use Psr\Container\ContainerInterface;
+use Symfony\Component\JsonEncoder\Attribute\Denormalizer;
+use Symfony\Component\JsonEncoder\Attribute\EncodedName;
+use Symfony\Component\JsonEncoder\Decode\Denormalizer\DenormalizerInterface;
+use Symfony\Component\JsonEncoder\Exception\InvalidArgumentException;
+use Symfony\Component\JsonEncoder\Mapping\PropertyMetadataLoaderInterface;
+use Symfony\Component\TypeInfo\TypeResolver\TypeResolverInterface;
+
+/**
+ * Enhances properties decoding metadata based on properties' attributes.
+ *
+ * @author Mathias Arlaud <mathias.arlaud@gmail.com>
+ *
+ * @internal
+ */
+final class AttributePropertyMetadataLoader implements PropertyMetadataLoaderInterface
+{
+    public function __construct(
+        private PropertyMetadataLoaderInterface $decorated,
+        private ContainerInterface $denormalizers,
+        private TypeResolverInterface $typeResolver,
+    ) {
+    }
+
+    public function load(string $className, array $options = [], array $context = []): array
+    {
+        $initialResult = $this->decorated->load($className, $options, $context);
+        $result = [];
+
+        foreach ($initialResult as $initialEncodedName => $initialMetadata) {
+            $attributesMetadata = $this->getPropertyAttributesMetadata(new \ReflectionProperty($className, $initialMetadata->getName()));
+            $encodedName = $attributesMetadata['name'] ?? $initialEncodedName;
+
+            if (null === $denormalizer = $attributesMetadata['denormalizer'] ?? null) {
+                $result[$encodedName] = $initialMetadata;
+
+                continue;
+            }
+
+            if (\is_string($denormalizer)) {
+                $denormalizerService = $this->getAndValidateDenormalizerService($denormalizer);
+                $normalizedType = $denormalizerService::getNormalizedType();
+
+                $result[$encodedName] = $initialMetadata
+                    ->withType($normalizedType)
+                    ->withAdditionalDenormalizer($denormalizer);
+
+                continue;
+            }
+
+            $normalizedType = $this->typeResolver->resolve((new \ReflectionFunction($denormalizer))->getParameters()[0]);
```

Comment:
what's happening if this throws a reflection exception? or if there are no parameters?